"consider using a semicolon here" when ; is present with dbg!() inside inspect()

````rust
fn main() {
let sum: i32 = [1,2,3].iter().inspect(|n| dbg!(n) ).sum();
}
````
This will suggest adding a `;`

````
error[E0308]: mismatched types
 --> src/main.rs:3:43
  |
3 | let sum: i32 = [1,2,3].iter().inspect(|n| dbg!(n) ).sum();
  |                                           ^^^^^^^
  |                                           |
  |                                           expected `()`, found `&&{integer}`
  |                                           expected this to be `()`
  |                                           help: consider using a semicolon here
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
 ````
 
 But doing so will "break" the code even more and not fix the error:
 
 ````
 error: expected one of `)`, `,`, `.`, `?`, or an operator, found `;`
 --> src/main.rs:2:50
  |
2 | let sum: i32 = [1,2,3].iter().inspect(|n| dbg!(n); ).sum();
  |                                                  ^ expected one of `)`, `,`, `.`, `?`, or an operator

error[E0308]: mismatched types
 --> src/main.rs:2:43
  |
2 | let sum: i32 = [1,2,3].iter().inspect(|n| dbg!(n); ).sum();
  |                                           ^^^^^^^
  |                                           |
  |                                           expected `()`, found `&&{integer}`
  |                                           expected this to be `()`
  |                                           help: consider using a semicolon here
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

````
Could rust suggest adding `{    }` around the macro?

The problem is we look at the macro-expanded code when generating this suggestion, which looks like this:

```rust
fn main() {
    let sum: i32 = [1, 2, 3]
        .iter()
        .inspect(|n| match n {
            tmp => {
                ... macro-generated stuff here ...
                tmp
            }
        })
        .sum();
}
```

Note that there are braces around `tmp` here. In this case adding a semicolon *does* work. So in that sense the suggestion is correct.

The problem is it's looking at macro-expanded code and making the suggestion based on that.

I can think of a few ways of improving this error:

- Not making any suggestions is better than making wrong suggestions, so we could just avoid this suggestion when the problematic code is macro expansion. If I implement this the error becomes:
  ```
  error[E0308]: mismatched types
   --> src/main.rs:2:49
    |
  2 |     let sum: i32 = [1, 2, 3].iter().inspect(|n| dbg!(n)).sum();
    |                                                 ^^^^^^^
    |                                                 |
    |                                                 expected `()`, found `&&{integer}`
    |                                                 expected this to be `()`
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
  ```

- Somehow check that the pre-macro-expansion code doesn't have `{}` but the expected type is `()`, and suggest adding braces. I don't know if it's possible to look at pre-macro-expansion code when type checking.

- Try to recover by adding a semicolon *in the original code* and type checking the macro expansion again. I'm not sure if this is even possible (possibly requires huge amounts of work unless the code is already there for this kind of recovery). If this is possible then we only suggest adding a semicolon when type checking the recovered version (with `;`) works.

--- a/compiler/rustc_typeck/src/check/coercion.rs
+++ b/compiler/rustc_typeck/src/check/coercion.rs
@@ -42,6 +42,7 @@ use rustc_hir as hir;
 use rustc_hir::def_id::DefId;
 use rustc_infer::infer::type_variable::{TypeVariableOrigin, TypeVariableOriginKind};
 use rustc_infer::infer::{Coercion, InferOk, InferResult};
+use rustc_middle::lint::in_external_macro;
 use rustc_middle::ty::adjustment::{
     Adjust, Adjustment, AllowTwoPhase, AutoBorrow, AutoBorrowMutability, PointerCast,
 };
@@ -1448,7 +1449,12 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
                 expected.is_unit(),
                 pointing_at_return_type,
             ) {
-                if cond_expr.span.desugaring_kind().is_none() {
+                // If the block is from an external macro, then do not suggest
+                // adding a semicolon, because there's nowhere to put it.
+                // See issue #81943.
+                if cond_expr.span.desugaring_kind().is_none()
+                    && !in_external_macro(fcx.tcx.sess, cond_expr.span)
+                {
                     err.span_label(cond_expr.span, "expected this to be `()`");
                     if expr.can_have_side_effects() {
                         fcx.suggest_semicolon_at_end(cond_expr.span, &mut err);

--- a/compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs
+++ b/compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs
@@ -10,6 +10,7 @@ use rustc_hir::def::{CtorOf, DefKind};
 use rustc_hir::lang_items::LangItem;
 use rustc_hir::{ExprKind, ItemKind, Node};
 use rustc_infer::infer;
+use rustc_middle::lint::in_external_macro;
 use rustc_middle::ty::{self, Ty};
 use rustc_span::symbol::kw;
 
@@ -44,7 +45,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         blk_id: hir::HirId,
     ) -> bool {
         let expr = expr.peel_drop_temps();
-        if expr.can_have_side_effects() {
+        // If the expression is from an external macro, then do not suggest
+        // adding a semicolon, because there's nowhere to put it.
+        // See issue #81943.
+        if expr.can_have_side_effects() && !in_external_macro(self.tcx.sess, cause_span) {
             self.suggest_missing_semicolon(err, expr, expected, cause_span);
         }
         let mut pointing_at_return_type = false;

--- a/src/test/ui/typeck/auxiliary/issue-81943-lib.rs
+++ b/src/test/ui/typeck/auxiliary/issue-81943-lib.rs
@@ -1,0 +1,7 @@
+pub fn g(t: i32) -> i32 { t }
+// This function imitates `dbg!` so that future changes
+// to its macro definition won't make this test a dud.
+#[macro_export]
+macro_rules! d {
+  ($e:expr) => { match $e { x => { $crate::g(x) } } }
+}

--- a/src/test/ui/typeck/issue-81943.rs
+++ b/src/test/ui/typeck/issue-81943.rs
@@ -1,0 +1,13 @@
+// aux-build:issue-81943-lib.rs
+extern crate issue_81943_lib as lib;
+
+fn f<F: Fn(i32)>(f: F) { f(0); }
+fn g(t: i32) -> i32 { t }
+fn main() {
+  f(|x| lib::d!(x)); //~ERROR
+  f(|x| match x { tmp => { g(tmp) } }); //~ERROR
+  macro_rules! d {
+    ($e:expr) => { match $e { x => { g(x) } } } //~ERROR
+  }
+  f(|x| d!(x));
+}

--- a/src/test/ui/typeck/issue-81943.stderr
+++ b/src/test/ui/typeck/issue-81943.stderr
@@ -1,0 +1,51 @@
+error[E0308]: mismatched types
+  --> $DIR/issue-81943.rs:7:9
+   |
+LL |   f(|x| lib::d!(x));
+   |         ^^^^^^^^^^ expected `()`, found `i32`
+   |
+   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error[E0308]: mismatched types
+  --> $DIR/issue-81943.rs:8:28
+   |
+LL |   f(|x| match x { tmp => { g(tmp) } });
+   |         -------------------^^^^^^----
+   |         |                  |
+   |         |                  expected `()`, found `i32`
+   |         expected this to be `()`
+   |
+help: consider using a semicolon here
+   |
+LL |   f(|x| match x { tmp => { g(tmp); } });
+   |                                  ^
+help: consider using a semicolon here
+   |
+LL |   f(|x| match x { tmp => { g(tmp) } };);
+   |                                      ^
+
+error[E0308]: mismatched types
+  --> $DIR/issue-81943.rs:10:38
+   |
+LL |     ($e:expr) => { match $e { x => { g(x) } } }
+   |                    ------------------^^^^----
+   |                    |                 |
+   |                    |                 expected `()`, found `i32`
+   |                    expected this to be `()`
+LL |   }
+LL |   f(|x| d!(x));
+   |         ----- in this macro invocation
+   |
+   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
+help: consider using a semicolon here
+   |
+LL |     ($e:expr) => { match $e { x => { g(x); } } }
+   |                                          ^
+help: consider using a semicolon here
+   |
+LL |     ($e:expr) => { match $e { x => { g(x) } }; }
+   |                                              ^
+
+error: aborting due to 3 previous errors
+
+For more information about this error, try `rustc --explain E0308`.

